src/main/java/com/connexta/search/index/IndexServiceImpl.java

```diff
@@ -0,0 +1,124 @@
+import org.springframework.web.reactive.function.client.WebClient;
+import reactor.core.publisher.Mono;
+
+@Slf4j
+@AllArgsConstructor
+public class IndexServiceImpl implements IndexService {
+
+  private static final String EXT_EXTRACTED_TEXT = "ext.extracted.text";
+
+  @NotNull private final CrudRepository<Index, String> crudRepository;
+  @NotNull private final WebClient webClient;
+
+  @Override
+  public void index(@NotBlank final String datasetId, @NotNull final URI irmUri)
+      throws IndexException {
+    // TODO check that the dataset exists in S3
+
+    final boolean idAlreadyExists;
+    try {
+      idAlreadyExists = crudRepository.existsById(datasetId);
+    } catch (final Exception e) {
+      throw new IndexException(
+          INTERNAL_SERVER_ERROR,
+          String.format("Unable to query index for datasetId=%s", datasetId),
+          e);
+    }
+    if (idAlreadyExists) {
+      throw new IndexException(
+          BAD_REQUEST,
+          String.format("datasetId=%s already exists. Overwriting is not supported", datasetId));
+    }
+
+    final Resource resource;
+    try {
+      final HttpStatus expectedHttpStatus = HttpStatus.OK;
+      resource =
+          webClient
+              .get()
+              .uri(irmUri)
+              .retrieve()
+              .onStatus(
+                  httpStatus -> !expectedHttpStatus.equals(httpStatus),
+                  clientResponse ->
+                      Mono.error(
+                          () ->
+                              new IndexException(
+                                  BAD_REQUEST,
+                                  String.format(
+                                      "Excepted %s but received status code of %s when getting resource at irmUri=%s",
+                                      expectedHttpStatus, clientResponse.statusCode(), irmUri))))
+              .bodyToMono(Resource.class)
+              .block();
+    } catch (final Exception e) {
+      throw new IndexException(
+          BAD_REQUEST, String.format("Unable to complete GET request to irmUri=%s", irmUri), e);
+    }
+
+    if (resource == null) {
+      throw new IndexException(
+          BAD_REQUEST, String.format("Unable to complete GET request to irmUri=%s", irmUri));
+    }
+
+    final String contents;
+    try (final InputStream inputStream = resource.getInputStream()) {
+      try {
+        // TODO Parse IRM instead of JSON CST
+        contents = getElement(parseJson(inputStream), EXT_EXTRACTED_TEXT);
+      } catch (IOException e) {
+        throw new IndexException(
+            INTERNAL_SERVER_ERROR,
+            String.format("Unable to convert InputStream to JSON for irmUri=%s", irmUri),
+            e);
+      }
+    } catch (IOException e) {
+      throw new IndexException(
+          BAD_REQUEST, String.format("Unable to get InputStream from irmUri=%s", irmUri), e);
+    }
+
+    log.info("Attempting to index datasetId={}", datasetId);
```

Comment:
Can probably dump this log and add the datasetId to the IndexException so it is captured for both paths.